### ISSUE TITLE
Reproducible sampling, reset Sampler

### ISSUE THREAD

**Issue by JavierEscobarOrtiz** [optuna/optuna#3526]
### Environment

- Optuna version: 2.10.0
- Python version: 3.8.0 (default, Nov  6 2019, 21:49:08) [GCC 7.3.0]
- OS: Linux-5.13.0-40-generic-x86_64-with-glibc2.10
- (Optional) Other libraries and their versions:

ipykernel           6.4.1
numpy               1.22.0
optuna              2.10.0
pandas              1.3.5
session_info        1.0.0
skforecast          0.5.dev1
sklearn             1.0.2
skopt               0.9.0

IPython             7.29.0
jupyter_client      7.1.0
jupyter_core        4.9.1
jupyterlab          3.2.5
notebook            6.4.6

Python 3.8.0 (default, Nov  6 2019, 21:49:08) [GCC 7.3.0]
Linux-5.13.0-40-generic-x86_64-with-glibc2.10

Session information updated at 2022-04-29 16:21

### Description

Hello,

I am passing some arguments to `optuna.create_study()` with a `dict`:

``` python
dict_to_create_study = {'sampler': TPESampler(seed=123)}
    
study = optuna.create_study(**dict_to_create_study)
```
Then, I am running different studies in a for loop. 

This code represents my problem:

``` python
import optuna
from optuna.samplers import TPESampler

def objective(trial):
    x = trial.suggest_float("x", -10, 10)
    return x ** 2

# dict_to_create_study is outside of the for loop
dict_to_create_study = {'sampler': TPESampler(seed=123)}

for i in range(2):
    
    study = optuna.create_study(**dict_to_create_study)
    study.optimize(objective, n_trials=10)
    
    params_list = []

    for trial in study.get_trials():
        params_list.append(trial.params)

    print(f'params_list_{i}:', params_list)
    print('')
```
The output are two different lists whereas I expect `params_list_0 == params_list_1`.
```
params_list_0: [{'x': 3.9293837119572324}, {'x': -4.277213300992411}, {'x': -5.462970928715938}, {'x': 1.0262953816578246}, {'x': 4.389379395711261}, {'x': -1.5378707975107808}, {'x': 9.61528396769231}, {'x': 3.6965947716972654}, {'x': -0.3813619703127813}, {'x': -2.1576496361169895}]

params_list_1: [{'x': -3.1364396769826115}, {'x': 4.580994147680833}, {'x': -1.2285551064075122}, {'x': -8.806442067808632}, {'x': -2.039114893391371}, {'x': 4.759908114640714}, {'x': -6.35016539093}, {'x': -6.49096487705015}, {'x': 0.631027476836767}, {'x': 0.6365517419373212}]
```

Since I have to declare my `dict` outside the for loop, is there any way for the `Sampler` to give the same values?

In my opinion, the `Sampler` is not restarted after each iteration of the for loop. Because, if I run the code as below, it gives me the same results:

``` python
import optuna
from optuna.samplers import TPESampler

def objective(trial):
    x = trial.suggest_float("x", -10, 10)
    return x ** 2

for i in range(2):
    
    # Now dict_to_create_study is inside of the for loop
    dict_to_create_study = {'sampler': TPESampler(seed=123)}
    
    study = optuna.create_study(**dict_to_create_study)
    study.optimize(objective, n_trials=10)
    
    params_list = []

    for trial in study.get_trials():
        params_list.append(trial.params)

    print(f'params_list_{i}:', params_list)
    print('')
```
```
params_list_0: [{'x': 3.9293837119572324}, {'x': -4.277213300992411}, {'x': -5.462970928715938}, {'x': 1.0262953816578246}, {'x': 4.389379395711261}, {'x': -1.5378707975107808}, {'x': 9.61528396769231}, {'x': 3.6965947716972654}, {'x': -0.3813619703127813}, {'x': -2.1576496361169895}]

params_list_1: [{'x': 3.9293837119572324}, {'x': -4.277213300992411}, {'x': -5.462970928715938}, {'x': 1.0262953816578246}, {'x': 4.389379395711261}, {'x': -1.5378707975107808}, {'x': 9.61528396769231}, {'x': 3.6965947716972654}, {'x': -0.3813619703127813}, {'x': -2.1576496361169895}]
```

Thank you very much!

### Error messages, stack traces, or logs

```shell
Not reproducible
```

### Steps to reproduce

```python
import optuna
from optuna.samplers import TPESampler

def objective(trial):
    x = trial.suggest_float("x", -10, 10)
    return x ** 2

dict_to_create_study = {'sampler': TPESampler(seed=123)}

for i in range(2):
    
    study = optuna.create_study(**dict_to_create_study)
    study.optimize(objective, n_trials=10)
    
    params_list = []

    for trial in study.get_trials():
        params_list.append(trial.params)

    print(f'params_list_{i}:', params_list)
    print('')
```

**Comment by JavierEscobarOrtiz** [comment]
Hello @nzw0301,

Thank you very much for your answer!

In my real problem, each loop iteration trains a different model and I would like to have the same starting point in all of them. 

If I don't misunderstand your `study.optimize` in a loop where `model_iteraion_1 == `model_iteration_2` the `TPESampler` will find the same parameters to try. ¿Is it correct?

Thank you very much again! I am using your library to optimize forecasting problems in [skforecast](https://github.com/JoaquinAmatRodrigo/skforecast)

**Comment by nzw0301** [comment]
We are glad to hear Optuna is useful for your project!

> In my real problem, each loop iteration trains a different model and I would like to have the same starting point in all of them.

So in my understanding, we do not need to reuse the same TPE sampler instance among loops; instantiating the TPE sampler in the loop is enough to do so.

> If I don't misunderstand your study.optimize in a loop where model_iteraion_1 == model_iteration_2theTPESampler` will find the same parameters to try. ¿Is it correct?

If models are different or their hyperparameter spaces are different, it changes. At least, I suppose the returned objective values are different among different models (or let's say different loops). In this case, since the TPE sampler suggests hyperparameters based on the finished trials, the suggested hyperparameters are also different even though we use the same random seed for the TPE sampler.

**Comment by JavierEscobarOrtiz** [comment]
Hello @nzw0301, thank you very much for sharing yogur thoughts with me.

All good! I close the issue 👍
